common/client/multi_node_test.go

```diff
@@ -0,0 +1,629 @@
+		expectedError := errors.New("rpc failed to do the batch call")
+		rpc.On("BatchCallContext", mock.Anything, mock.Anything).Return(expectedError).Once()
+		node := newMockNode[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+		node.On("RPC").Return(rpc)
+		nodeSelector := newMockNodeSelector[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+		nodeSelector.On("Select").Return(node).Once()
+		mn := newTestMultiNode(t, multiNodeOpts{
+			selectionMode: NodeSelectionModeRoundRobin,
+			chainID:       chainID,
+		})
+		mn.nodeSelector = nodeSelector
+		err := mn.BatchCallContextAll(tests.Context(t), nil)
+		require.EqualError(t, err, expectedError.Error())
+	})
+	t.Run("Waits for all nodes to complete the call and logs results", func(t *testing.T) {
+		// setup RPCs
+		failedRPC := newMultiNodeRPCClient(t)
+		failedRPC.On("BatchCallContext", mock.Anything, mock.Anything).
+			Return(errors.New("rpc failed to do the batch call")).Once()
+		okRPC := newMultiNodeRPCClient(t)
+		okRPC.On("BatchCallContext", mock.Anything, mock.Anything).Return(nil).Twice()
+
+		// setup ok and failed auxiliary nodes
+		okNode := newMockSendOnlyNode[types.ID, multiNodeRPCClient](t)
+		okNode.On("RPC").Return(okRPC).Once()
+		failedNode := newMockNode[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+		failedNode.On("RPC").Return(failedRPC).Once()
+
+		// setup main node
+		mainNode := newMockNode[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+		mainNode.On("RPC").Return(okRPC)
+		nodeSelector := newMockNodeSelector[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+		nodeSelector.On("Select").Return(mainNode).Once()
+		lggr, observedLogs := logger.TestLoggerObserved(t, zap.DebugLevel)
+		mn := newTestMultiNode(t, multiNodeOpts{
+			selectionMode: NodeSelectionModeRoundRobin,
+			chainID:       types.RandomID(),
+			nodes:         []Node[types.ID, types.Head[Hashable], multiNodeRPCClient]{failedNode, mainNode},
+			sendonlys:     []SendOnlyNode[types.ID, multiNodeRPCClient]{okNode},
+			logger:        lggr,
+		})
+		mn.nodeSelector = nodeSelector
+
+		err := mn.BatchCallContextAll(tests.Context(t), nil)
+		require.NoError(t, err)
+		tests.RequireLogMessage(t, observedLogs, "Secondary node BatchCallContext failed")
+	})
+}
+
+func TestMultiNode_SendTransaction(t *testing.T) {
+	t.Parallel()
+	t.Run("Fails if failed to select active node", func(t *testing.T) {
+		chainID := types.RandomID()
+		mn := newTestMultiNode(t, multiNodeOpts{
+			selectionMode: NodeSelectionModeRoundRobin,
+			chainID:       chainID,
+		})
+		nodeSelector := newMockNodeSelector[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+		nodeSelector.On("Select").Return(nil).Once()
+		nodeSelector.On("Name").Return("MockedNodeSelector").Once()
+		mn.nodeSelector = nodeSelector
+		err := mn.SendTransaction(tests.Context(t), nil)
+		require.EqualError(t, err, ErroringNodeError.Error())
+	})
+	t.Run("Returns response of active node", func(t *testing.T) {
```

Comment:
This message describes a happy path test case. Here we select a node, but the RPC fails. Perhaps a more fitting name would be:
```suggestion
	t.Run("Returns error if RPC call fails", func(t *testing.T) {
```